docs/rfc/rfc-006-event-subscription.md

```diff
@@ -0,0 +1,189 @@
+  stability reasons (e.g., `tx_search`). A node may still need to expose
+  events, however, to support UI tools.
+
+  This matters, because all the methods share the same network endpoint. While
+  it is possible to block the top-level GET and POST handlers with a proxy,
+  exposing the `/websocket` handler exposes not _only_ the event subscription
+  methods, but the rest of the service as well.
+
+### Possible Improvements
+
+There are several things we could do to improve the experience of developers
+who need to subscribe to events from the consensus node. These are not all
+mutually exclusive.
+
+1. **Split event subscription into a separate service**. Instead of exposing
+   event subscription on the same endpoint as the rest of the RPC service,
+   dedicate a separate endpoint on the node for _only_ event subscription.  The
+   rest of the RPC services (_sans_ events) would remain as-is.
+
+   This would make it easy to disable or firewall outside access to sensitive
+   RPC methods, without blocking access to event subscription (and vice versa).
+
+2. **Use a different protocol for event subscription.** There are various ways
+   we could approach this, depending how much we're willing to shake up the
+   current API. Here are sketches of a few options:
+
+   - Keep the websocket, but rework the API to be more JSON-RPC compliant,
+     perhaps by converting event delivery into notifications.  This is less
+     up-front change for existing clients, but retains all of the existing
+     implementation complexity, and doesn't contribute much toward more serious
+     performance and UX improvements later.
+
+   - Switch from websocket to plain HTTP, and rework the subscription API to
+     use a more conventional request/response pattern instead of streaming.
+     This is a little more up-front work for existing clients, but leverages
+     better library support for clients not written in Go. The protocol would
+     be more chatty, but we could mitigate that with batching.
+
+	 This has a nice incidental benefit that it would become easier to debug
+     subscriptions from the command-line, like the other RPC methods.
+
+   - Switch to gRPC: Preserves a persistent connection and gives us a more
+     efficient binary wire format (protobuf), at the cost of much more work for
+     clients and harder debugging. This may be the best option if performance
+     and server load are our top concerns. (Given that we are currently using
+     JSON-RPC, however, I'm not convinced throughput is the bottleneck here).
+
+3. **Delegate event subscriptions to a proxy.** Give responsibility for
+     managing event subscription to a proxy that runs separately from the node,
+     and switch the node to push events to the proxy (like a webhook) instead
+     of serving subscribers directly.  This is more work for the operator
+     (another process to configure and run) but may scale better for big
+     networks.
```

Comment:
That is true, though not more so than the way it works now. Switching pull to push doesn't fix the possibility that some messages won't be delivered, it's more about making the load more predictable for the node.

Right now, the node winds up buffering multiple copies of events, for each subscriber. With a push-proxy, it would only need one copy at a time, and could do them in batches when other activity is lower.

Reply:
proxy as in "caller's service" or as in something that we'd have to manage? 

I like the idea of the deduplication and the predictable workload.

Reply:
I'm not super wild about having additional processes to manage subscriptions: it just sort of feels like moving the food around on our plate as much as anything, if we have to provide both ends of the pipeline: though making subscription handling their responsibility (either webhooks, or pushing to kafka or whatever.) seem fine. 

Having said that, I guess I was sort of always in the camp of the cursor-like options anyway, if we expect to find ourselves in the business of having to support this long term (it has other benefits that you point out in that section too, particularly around debugability.)

Reply:
I agree, and my intuition is that if we're going to do something like this, we should probably go further (e.g., option (4) below).  I wanted to document the option, but I'm not strongly advocating for it.

The one way in which this isn't just rearranging deck chairs is that it makes integration with a third-party tool easier. But that's only useful if scalability is a primary concern, and I don't know whether it is.

Reply:
Makes sense. 

I'd be ok having some of the discussion about our preferences and opinions captured in the document/discussion section, just so we don't have to re-litigate/replay consensus even if it's only in our minds, because I think we're of a similar mind.